internal/agent/hooks/kcrypt_uki.go

```diff
@@ -96,10 +97,40 @@ func (k KcryptUKI) Run(c config.Config, _ v1.Spec) error {
 		c.Logger.Infof("Done encrypting %s", p)
 	}
 
+	_, _ = utils.SH("sync")
+
 	err = kcrypt.UnlockAll(true)
 	if err != nil {
 		return err
 	}
+	// Close the unlocked partitions after dealing with them, otherwise we leave them open and they can be mounted by anyone
+	defer func() {
+		for _, p := range append([]string{"COS_OEM", "COS_PERSISTENT"}, c.Install.Encrypt...) {
+			part, _, err := kcrypt.FindPartition(p)
+			if err != nil {
+				continue
+			}
+			c.Logger.Debugf("Closing /dev/mapper/%s", part)
+			out, err := utils.SH(fmt.Sprintf("cryptsetup close /dev/mapper/%s", part))
+			c.Logger.Debugf("cryptsetup close /dev/mapper/%s: %s", part, out)
+			if err != nil {
+				c.Logger.Errorf("could not close /dev/mapper/%s: %s", part, out)
+			}
+		}
+	}()
+
+	// Here it can take the oem partition a bit of time to appear after unlocking so we need to retry a couple of time with some waiting
+	for i := 0; i < 10; i++ {
+		c.Logger.Infof("Waiting for unlocked partition to appear")
+		part, _ := utils.SH("blkid -L COS_OEM")
+		if part == "" {
+			c.Logger.Infof("Partition not found, waiting 1 second before retrying")
+			time.Sleep(5 * time.Second)
```

Comment:
how about an incremental wait? so it does something like 1 sec, 5 secs, 15 secs or so ?

Reply:
good point!